lib/fog/libvirt/requests/compute/list_domains.rb

```diff
@@ -46,6 +46,24 @@ def boot_order xml
           xml_elements(xml, "domain/os/boot", "dev")
         end
 
+        def firmware(xml)
+          firmware_from_loader = xml_elements(xml, "domain/os/loader", "type").first
+
+          case firmware_from_loader
+          when 'pflash'
+            'efi'
+          when 'rom'
+            'bios'
+          else
+            xml_elements(xml, "domain/os", "firmware").first || 'bios'
+          end
+        end
+
+        # we rely on the fact that the secure attribute is only present when secure boot is enabled
+        def secure_boot_enabled?(xml)
+          xml_elements(xml, "domain/os/loader", "secure").first == 'yes'
```

Comment:
Untested, but should this be:
```suggestion
          xml_element(xml, "domain/os/loader", "secure") == 'yes'
```

Reply:
`xml_elements(xml, "domain/os/loader", "secure")` returns an array, and the same applies to all other suggestions as well.

Reply:
But that's why I'm suggesting `xml_element` (singular). That code already has a `.first`:
https://github.com/fog/fog-libvirt/blob/11f3821777becf14b3bb3c14122f8cb5af55a8c7/lib/fog/libvirt/models/compute/util/util.rb#L9-L12

Reply:
Oh sorry, I missed that change. Checking.

Reply:
The problem is that in some cases the xml element (`xml/path`) is empty, resulting in an empty array. In such cases, if we use `xml_element` we'd get an error for trying to call `.first[attribute]` on nil. 
To address this, we can either modify the `xml_element` method to handle this issue or continue using `xml_elements` instead.

Reply:
I was afraid of that because I already noticed it when looking at the method. Short term I'd lean to accepting `xml_elements`. Changing `xml_element` to silently return `nil` if the element doesn't exist could break existing code (because it wants to distinguish between the element not existing vs the attribute not existing) and I'd rather not block on figuring that out.

Having said that, I wonder why the code is parsing the XML all the time. Note every `xml_element` and `xml_elements` call parses the XML which probably makes it much slower than it needs to be.

Not a blocker to merging this, but hints for a follow up. I think if `domain_to_attributes` parses the XML and then passes on the parsed tree this would also work:
```suggestion
          loader = (xml / "domain/os/loader")&.first
          loader && loader[:secure] == 'yes'
```

That means you only parse the XML once instead of 6 or so times. Then I think the code is also simple enough to not need `xml_element` or `xml_elements`.

Reply:
I agree that we can improve this, but I'd suggest opening a separate PR for it since the change also affects existing code, not just the part I added. To implement your suggestion, we'd still need to parse the XML, but we could reduce the number of times we do it.

Would it be acceptable to merge this PR as is and then open a new PR to handle the refactoring?

Reply:
We can resolve the XML parsing later, but I'd still like to see an answer to https://github.com/fog/fog-libvirt/pull/155#discussion_r1825764058.